cpp/lib/ClientConfigParser.cpp

```diff
@@ -0,0 +1,315 @@
+    derivedConfigPath = clientConfigEnv;
+    CXX_LOG_INFO("sf", "ClientConfigParser", "loadClientConfig",
+      "Using client configuration path from an environment variable: %s", clientConfigEnv);
+  }
+  else
+  {
+    // 3. Try DLL binary dir
+    std::string binaryDir = getBinaryPath();
+    std::string binaryDirFilePath = binaryDir + SF_CLIENT_CONFIG_FILE_NAME;
+    if (checkFileExists(binaryDirFilePath))
+    {
+      derivedConfigPath = binaryDirFilePath;
+      CXX_LOG_INFO("sf", "ClientConfigParser", "loadClientConfig",
+        "Using client configuration path from binary directory: %s", binaryDirFilePath.c_str());
+    }
+    else
+    {
+#if defined(WIN32) || defined(_WIN64)
+      // 4. Try user home dir
+      std::string homeDir = sf_getenv_s("USERPROFILE", envbuf, sizeof(envbuf));
+      if (homeDir.empty())
+      {
+        // USERPROFILE is empty, try HOMEDRIVE and HOMEPATH
+        std::string homeDriveEnv = sf_getenv_s("HOMEDRIVE", envbuf, sizeof(envbuf));
+        std::string homePathEnv = sf_getenv_s("HOMEPATH", envbuf, sizeof(envbuf));
+        if (!homeDriveEnv.empty() && !homePathEnv.empty())
+        {
+          homeDir = homeDriveEnv + homePathEnv;
+        }
+      }
+      std::string homeDirFilePath = homeDir + PATH_SEP + SF_CLIENT_CONFIG_FILE_NAME;
+      if (checkFileExists(homeDirFilePath))
+      {
+        derivedConfigPath = homeDirFilePath;
+        CXX_LOG_INFO("sf", "ClientConfigParser", "loadClientConfig",
+          "Using client configuration path from home directory: %s", homeDirFilePath.c_str());
+      }
+#else
+      // 4. Try user home dir
+      if (const char* homeDir = sf_getenv_s("HOME", envbuf, sizeof(envbuf)))
+      {
+        std::string homeDirFilePath = std::string(homeDir) + PATH_SEP + SF_CLIENT_CONFIG_FILE_NAME;
+        if (checkFileExists(homeDirFilePath))
+        {
+          derivedConfigPath = homeDirFilePath;
+          CXX_LOG_INFO("sf", "ClientConfigParser", "loadClientConfig",
+            "Using client configuration path from home directory: %s", homeDirFilePath.c_str());
+        }
+      }
+#endif
+    }
+  }
+  if (!derivedConfigPath.empty())
+  {
+    parseConfigFile(derivedConfigPath, out_clientConfig);
+  }
+}
+
+// Private =====================================================================
+////////////////////////////////////////////////////////////////////////////////
+bool ClientConfigParser::checkFileExists(const std::string& in_filePath)
+{
+  FILE* file = sf_fopen(&file, in_filePath.c_str(), "r");
+  if (file != nullptr)
+  {
+    fclose(file);
+    return true;
+  }
+  return false;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+void ClientConfigParser::parseConfigFile(
+  const std::string& in_filePath,
+  ClientConfig& out_clientConfig)
+{
+  cJSON* jsonConfig;
+  FILE* configFile;
+  try 
+  {
+    configFile = sf_fopen(&configFile, in_filePath.c_str(), "r");
```

Comment:
Unfortunately std::filesystem isn't in osx10.14 yet so it can't be used and there seems to be a 32-bit windows debug issue with reading file with fstream

Reply:
That means we cannot use std::fstream at all? Can you show me the assertion?

Reply:
Not exactly, it only happens for Windows 32-bit built with MT. All other configurations seem to work fine. If you prefer, I can ifdef it to not run when it's windows debug and it should work fine. The assertion says something like this: 
Debug Assertion Failed!

File: minkernel\crts\ucrt\src\appcrt\lowio\read.cpp

Expression: _osfile(fh) & FOPEN

Reply:
I'm still running into this issue with the tests run on jenkins. Basically the test run just hangs after it hits the assertion error. In the mean time, I've added an ifdef to to run easy logging when in 32-bit debug. I've locally tested running on /MDd and it runs fine, but currently I think we build with dynamic runtime set to OFF